zkevm-circuits/src/evm_circuit/execution/mul_div_mod_shl_shr.rs

```diff
@@ -0,0 +1,351 @@
+        // - for `MOD`, two pops are dividend and divisor, and push is remainder.
+        // - for `SHL`, two pops are shift and quotient, and push is dividend.
+        // - for `SHR`, two pops are shift and dividend, and push is quotient.
+        cb.stack_pop(pop1.expr());
+        cb.stack_pop(
+            (is_mul.clone() + is_div.clone() + is_mod.clone()) * divisor.expr()
+                + is_shl.clone() * quotient.expr()
+                + is_shr.clone() * dividend.expr(),
+        );
+        cb.stack_push(
+            (is_mul.clone() + is_shl.clone()) * dividend.expr()
+                + (is_div.clone() + is_shr.clone())
+                    * quotient.expr()
+                    * (1.expr() - divisor_is_zero.expr())
+                + is_mod.clone() * remainder.expr() * (1.expr() - divisor_is_zero.expr()),
+        );
+
+        cb.require_zero(
+            "remainder < divisor when divisor != 0",
+            (1.expr() - divisor_is_zero.expr()) * (1.expr() - remainder_lt_divisor.expr()),
+        );
+
+        cb.require_zero(
+            "remainder == 0 for both opcode MUL and SHL",
+            (is_mul.clone() + is_shl.clone()) * (1.expr() - remainder_is_zero.expr()),
+        );
+
+        cb.require_zero(
+            "overflow == 0 for opcode DIV, MOD and SHR",
+            (is_div.clone() + is_mod.clone() + is_shr.clone()) * mul_add_words.overflow(),
+        );
+
+        let gas_cost = is_mul * OpcodeId::MUL.constant_gas_cost().expr()
+            + is_div * OpcodeId::DIV.constant_gas_cost().expr()
+            + is_mod * OpcodeId::MOD.constant_gas_cost().expr()
+            + is_shl * OpcodeId::SHL.constant_gas_cost().expr()
+            + is_shr * OpcodeId::SHR.constant_gas_cost().expr();
+
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(3.expr()),
+            program_counter: Delta(1.expr()),
+            stack_pointer: Delta(1.expr()),
+            gas_left: Delta(-gas_cost),
+            ..Default::default()
+        };
+
+        let same_context = SameContextGadget::construct(cb, opcode, step_state_transition);
+
+        Self {
+            same_context,
+            words: [quotient, divisor, remainder, dividend, pop1],
+            mul_add_words,
+            divisor_is_zero,
+            remainder_is_zero,
+            remainder_lt_divisor,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+        let indices = [step.rw_indices[0], step.rw_indices[1], step.rw_indices[2]];
+        let [pop1, pop2, push] = indices.map(|idx| block.rws[idx].stack_value());
+
+        // Shift values are only used for opcode SHL and SHR.
+        let shf0 = pop1.to_le_bytes()[0];
+        let shf_lt256 = pop1
```

Comment:
in the assignment, this shouldn't be needed. you can just directly check  `if U256::from(shf0 as u64) == pop1 {...}`, like how on line 160, there's `if pop2.is_zero() {...}`